### PR TITLE
YARN EMR Runner Part 3 [90% done]

### PR DESCRIPTION
**Pull request by aaronabf**
Issues observed in testing:

Timeout when connection to cluster master
```
requests.exceptions.ConnectTimeout: HTTPConnectionPool(host='ip-10-0-28-155.ec2.internal', port=8088): Max retries exceeded with url: /ws/v1/cluster/metrics (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f3e6c1b9b50>, 'Connection to ip-10-0-28-155.ec2.internal timed out. (connect timeout=10)'))
```

### CODE REVIEW

File: mrjob/yarnemr.py [hunk 1]
```diff
@@ -25,3 +64,243 @@ class YarnEMRJobRunner(EMRJobRunner):
...
+        """:py:class:`~mrjob.yarnemr.YarnEMRJobRunner` takes the same arguments
+        as :py:class:`~mrjob.emr.EMRJobRunner` with just one additional param,
+        `expected_memory`."""
+        super(YarnEMRJobRunner, self).__init__(**kwargs)
+
+        # TODO (aaron): raise something better than `Exception`
+
+        # arg required
+        if self._opts['expected_memory'] is None:
+            raise Exception
```
Comment (gregsterin) [inline mrjob/yarnemr.py]: Add a message along with the error.

File: mrjob/yarnemr.py [hunk 2]
```diff
@@ -25,3 +64,243 @@ class YarnEMRJobRunner(EMRJobRunner):
...
+
+        # arg required for ssh, no need to validate `ssh_bin` as there is a default
+        if self._opts['ec2_key_pair_file'] is None
+            raise Exception
```
Comment (gregsterin) [inline mrjob/yarnemr.py]: Add a message along with the error.

File: mrjob/yarnemr.py [hunk 3]
```diff
@@ -25,3 +64,243 @@ class YarnEMRJobRunner(EMRJobRunner):
...
+        function was renamed to `_launch_yarnemr_job` to avoid confusion.
+        """
+        self._create_s3_tmp_bucket_if_needed()
+        emr_client = self.make_emr_client()
+
+        # try to find a cluster from the pool
+        if (self._opts['pool_clusters'] and not self._cluster_id):
+            requested_mb = self._opts['expected_memory']
+            resource_constraints = [
+                _ResourceConstraint('availableVirtualCores', 2, operator.ge),
```
Comment (gregsterin) [inline mrjob/yarnemr.py]: Why 2? Ideally we provide the number of cores the job needs as well?

**Review by gregsterin — COMMENTED** [review]
Left some comments. Looking good overall.

File: mrjob/yarnemr.py [hunk 4]
```diff
@@ -25,3 +64,243 @@ class YarnEMRJobRunner(EMRJobRunner):
...
+                if not self._compare_cluster_setup(
+                        emr_client, cluster, req_pool_hash):
+                    invalid_clusters.append(cluster_id)
+                    return
+                valid_clusters.append(cluster_id)
+
+            available_md = self._check_cluster_state(cluster_host,
+                                                     resource_constraints)
+            if available_md != -1:
+                cluster_list.append(available_md, cluster_id)
```
Comment (gregsterin) [inline mrjob/yarnemr.py]: Syntax seems invalid.

```
>>> foo = []
>>> foo.append('a',  'b')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: append() takes exactly one argument (2 given)
```

Comment (aaronabf) [reply mrjob/yarnemr.py]: Fixed. Thanks for the catch.

**Comment by aaronabf** [comment]
Closing in favor of https://github.com/Affirm/mrjob/pull/11

**Pull request closed without merging**